arangod/Cache/Cache.cpp

```diff
@@ -431,46 +435,62 @@ bool Cache::migrate(std::shared_ptr<Table> newTable) {
-  TRI_ASSERT(table != nullptr);
-  std::shared_ptr<Table> oldAuxiliary = table->setAuxiliary(newTable);
-  TRI_ASSERT(oldAuxiliary == nullptr);
+  try {
+    newTable->setTypeSpecifics(_bucketClearer, _slotsPerBucket);
+    newTable->enable();
 
-  // do the actual migration
-  for (std::uint64_t i = 0; i < table->size();
-       i++) {  // need uint64 for end condition
-    migrateBucket(table->primaryBucket(static_cast<uint32_t>(i)),
-                  table->auxiliaryBuckets(static_cast<uint32_t>(i)), *newTable);
-  }
+    std::shared_ptr<cache::Table> table = this->table();
+    TRI_ASSERT(table != nullptr);
+    std::shared_ptr<Table> oldAuxiliary = table->setAuxiliary(newTable);
+    TRI_ASSERT(oldAuxiliary == nullptr);
+
+    // do the actual migration
+    for (std::uint64_t i = 0; i < table->size();
+         i++) {  // need uint64 for end condition
+      migrateBucket(table->primaryBucket(static_cast<uint32_t>(i)),
+                    table->auxiliaryBuckets(static_cast<uint32_t>(i)),
+                    *newTable);
+    }
 
-  // swap tables
-  std::shared_ptr<Table> oldTable;
-  {
-    SpinLocker taskGuard(SpinLocker::Mode::Write, _taskLock);
-    oldTable = this->table();
-    std::atomic_store_explicit(&_table, newTable, std::memory_order_release);
-    oldTable->setAuxiliary(std::shared_ptr<Table>());
-  }
+    // swap tables
+    std::shared_ptr<Table> oldTable;
+    {
+      SpinLocker taskGuard(SpinLocker::Mode::Write, _taskLock);
+      oldTable = this->table();
+      std::atomic_store_explicit(&_table, newTable, std::memory_order_release);
+      oldTable->setAuxiliary(std::shared_ptr<Table>());
+    }
 
-  TRI_ASSERT(oldTable != nullptr);
+    TRI_ASSERT(oldTable != nullptr);
 
-  // unmarking migrating flag
-  {
-    SpinLocker metaGuard(SpinLocker::Mode::Write, _metadata.lock());
-    _metadata.changeTable(newTable->memoryUsage());
-    TRI_ASSERT(_metadata.isMigrating());
-    _metadata.toggleMigrating();
-    TRI_ASSERT(!_metadata.isMigrating());
-  }
+    // unmarking migrating flag
+    {
+      SpinLocker metaGuard(SpinLocker::Mode::Write, _metadata.lock());
+      _metadata.changeTable(newTable->memoryUsage());
+      TRI_ASSERT(_metadata.isMigrating());
+      _metadata.toggleMigrating();
+      TRI_ASSERT(!_metadata.isMigrating());
+    }
 
-  // clear out old table and release it
-  oldTable->clear();
-  _manager->reclaimTable(std::move(oldTable), false);
+    clearMigrateFlag = false;
 
-  return true;
+    // clear out old table and release it
+    oldTable->clear();
+    _manager->reclaimTable(std::move(oldTable), false);
+
+    return true;
+  } catch (...) {
+    if (clearMigrateFlag) {
+      // we must be sure to have cleared the isMigrating() flag here
+      SpinLocker metaGuard(SpinLocker::Mode::Write, _metadata.lock());
+      TRI_ASSERT(_metadata.isMigrating());
+      _metadata.toggleMigrating();
+      TRI_ASSERT(!_metadata.isMigrating());
+    }
+    throw;
```

Comment:
absl::Cleanup (has cancel) or basics::ScopeGuard (not sure but probably also should has cal)